namesys/interface.go

```diff
@@ -60,39 +57,152 @@ type NameSystem interface {
+	Resolve(ctx context.Context, name string, options ...ResolveOption) (value path.Path, ttl time.Duration, err error)
+
+	// ResolveAsync performs recursive name lookup, like Resolve, but it returns entries as
+	// they are discovered in the DHT. Each returned result is guaranteed to be "better"
+	// (which usually means newer) than the previous one.
+	ResolveAsync(ctx context.Context, name string, options ...ResolveOption) <-chan ResolveResult
+}
+
+// ResolveOptions specifies options for resolving an IPNS Path.
+type ResolveOptions struct {
+	// Depth is the recursion depth limit.
+	Depth uint
+
+	// DhtRecordCount is the number of IPNS Records to retrieve from the DHT
+	// (the best record is selected from this set).
+	DhtRecordCount uint
 
-	// ResolveAsync performs recursive name lookup, like Resolve, but it returns
-	// entries as they are discovered in the DHT. Each returned result is guaranteed
-	// to be "better" (which usually means newer) than the previous one.
-	ResolveAsync(ctx context.Context, name string, options ...opts.ResolveOpt) <-chan Result
+	// DhtTimeout is the amount of time to wait for DHT records to be fetched
+	// and verified. A zero value indicates that there is no explicit timeout
+	// (although there is an implicit timeout due to dial timeouts within the DHT).
+	DhtTimeout time.Duration
+}
+
+// DefaultResolveOptions returns the default options for resolving an IPNS Path.
+func DefaultResolveOptions() ResolveOptions {
+	return ResolveOptions{
+		Depth:          DefaultDepthLimit,
+		DhtRecordCount: 16,
+		DhtTimeout:     time.Minute,
+	}
+}
+
+// ResolveOption is used to set a resolve option.
+type ResolveOption func(*ResolveOptions)
+
+// ResolveWithDepth sets [ResolveOptions.Depth].
+func ResolveWithDepth(depth uint) ResolveOption {
+	return func(o *ResolveOptions) {
+		o.Depth = depth
+	}
+}
+
+// ResolveWithDhtRecordCount sets [ResolveOptions.DhtRecordCount].
+func ResolveWithDhtRecordCount(count uint) ResolveOption {
+	return func(o *ResolveOptions) {
+		o.DhtRecordCount = count
+	}
+}
+
+// ResolveWithDhtTimeout sets [ResolveOptions.ResolveWithDhtTimeout].
+func ResolveWithDhtTimeout(timeout time.Duration) ResolveOption {
+	return func(o *ResolveOptions) {
+		o.DhtTimeout = timeout
+	}
+}
+
+// ProcessResolveOptions converts an array of [ResolveOption] into a [ResolveOptions] object.
+func ProcessResolveOptions(opts []ResolveOption) ResolveOptions {
+	resolveOptions := DefaultResolveOptions()
+	for _, option := range opts {
+		option(&resolveOptions)
+	}
+	return resolveOptions
 }
 
 // Publisher is an object capable of publishing particular names.
 type Publisher interface {
 	// Publish establishes a name-value mapping.
 	// TODO make this not PrivKey specific.
-	Publish(ctx context.Context, name ci.PrivKey, value path.Path, options ...opts.PublishOption) error
+	Publish(ctx context.Context, name ci.PrivKey, value path.Path, options ...PublishOption) error
```

Comment:
Similar to https://github.com/ipfs/boxo/pull/459/files#r1315812409, also potentially solve this `TODO`.

Reply:
I guess we could remove TODO by replacing `name ci.PrivKey` with  `name string` and a separate `signingKey ci.PrivKey`:

```go
Publish(ctx context.Context, name string, signingKey ci.PrivKey, value path.Path, options ...PublishOption) error
```

This would make API less error prone by requiring user to pass both IPNS name and private key (implementation could check if privkey matches expected name).
Since `name` is a string, this new `Publish` could be used for signing other things than IPNS.

Anything more (migrations etc)  feels out of the scope of his PR. The additional work to refactor things to generalize this interface does not feel worth the effort any time soon. `Publisher` is IPNS centric because there is no other use case than DNSPublisher, which itself is not feasible:

DNSLink does not provide end-to-end integrity; there is no generally accepted DNS protocol for making DNS updates that "just works" across providers, nor there is user need for having it built-in.  The way DNSLink ended up being used in real world   is setting record once, pointing it at `dnslink=/ipns/id` and then using IPNS for publishing updates. This is what ENS users do, since every DNS TXT record update on ENS costs gas,  DNSLink pointing at IPNS saves them money.

Reply:
Then I think I'll just remove the TODO, but not change it.